### PR TITLE
feat: Add support for mute_time_intervals, target_matchers and source_matchers Alertmanager config

### PR DESCRIPTION
**Pull request by Jammicus**
**NOTE:**

Going to be splitting this into two separate PRs to make things easier to review/tidy up commits, these new PRs are:

* Target matchers and source matchers support: https://github.com/prometheus-operator/prometheus-operator/pull/4175
* Mute time intervals: https://github.com/prometheus-operator/prometheus-operator/pull/4176

-----------

This PR is to add additional config options which Alertmanager introduced in v0.22.0. First time doing a large PR so if i've missed anything/made any mistakes please let me know! 

Addresses the following issues:

- https://github.com/prometheus-operator/prometheus-operator/issues/4087
- https://github.com/prometheus-operator/prometheus-operator/issues/4078

Note: I have bumped the default Alertmanager version to 0.22.0 as testing is dependent on that version

Example AlertmanagerConfig:

```
apiVersion: monitoring.coreos.com/v1alpha1
kind: AlertmanagerConfig
metadata:
  name: config-example
  labels:
    alertmanagerConfig: example
spec:
  route:
    groupBy: ['job']
    groupWait: 30s
    groupInterval: 5m
    repeatInterval: 12h
    receiver: 'wechat-example'
  receivers:
  - name: 'wechat-example'
    wechatConfigs:
    - apiURL: 'http://wechatserver:8080/'
      corpID: 'wechat-corpid'
      apiSecret:
        name: 'wechat-config'
        key: 'apiSecret'
  muteTimeInterval:
  - name: my_mute_time
    timeInterval:
    - times:
      - startTime: "09:00"
        endTime: "17:00"
      weekdays: ['sunday:tuesday', saturday]
      daysOfMonth: ["1:5", '-3:-1']
      months: ["1:3", "5:8", "12"]
      years: ['2020:2022', "2030"]
  - name: my_mute_time_again
    timeInterval:
    - times:
      - startTime: "10:00"
        endTime: "17:30"
      weekdays: ['sunday:monday', saturday]
      daysOfMonth: ["2:7", '-15:-2']
      months: ["1:2", "5:6"]
      years: ['2020:2025']
  inhibitRules:  
  - sourceMatchers: [ severity="critical" ]
    targetMatchers: [ severity="warning" ]
    equal: [ alertname, cluster, service ]
```

Which translates into the following config file within the Alertmanager pod:

```
route:
  receiver: "null"
  routes:
  - receiver: default-config-example-wechat-example
    group_by:
    - job
    match:
      namespace: default
    continue: true
    group_wait: 30s
    group_interval: 5m
    repeat_interval: 12h
inhibit_rules:
- target_matchers:
  - namespace=default
  - severity="warning"
  source_matchers:
  - namespace=default
  - severity="critical"
  equal:
  - alertname
  - cluster
  - service
receivers:
- name: "null"
- name: default-config-example-wechat-example
  wechat_configs:
  - api_secret: |
      wechat-secret
    api_url: http://wechatserver:8080/
    corp_id: wechat-corpid
templates: []
mute_time_intervals:
- name: my_mute_time
  time_intervals:
  - times:
    - start_time: "09:00"
      end_time: "17:00"
    weekdays: ['sunday:tuesday', saturday]
    days_of_month: ["1:5", '-3:-1']
    months: ["1:3", "5:8", "12"]
    years: ['2020:2022', "2030"]
- name: my_mute_time_again
  time_intervals:
  - times:
    - start_time: "10:00"
      end_time: "17:30"
    weekdays: ['sunday:monday', saturday]
    days_of_month: ["2:7", '-15:-2']
    months: ["1:2", "5:6"]
    years: ['2020:2025']
```

CRD Explainations:

```
KIND:     AlertmanagerConfig
VERSION:  monitoring.coreos.com/v1alpha1

RESOURCE: inhibitRules <[]Object>

DESCRIPTION:
     List of inhibition rules. The rules will only apply to alerts matching the
     resource’s namespace.

     InhibitRule defines an inhibition rule that allows to mute alerts when
     other alerts are already firing. See
     https://prometheus.io/docs/alerting/latest/configuration/#inhibit_rule

FIELDS:
   equal	<[]string>
     Labels that must have an equal value in the source and target alert for the
     inhibition to take effect.

   sourceMatch	<[]Object>
     Matchers for which one or more alerts have to exist for the inhibition to
     take effect. The operator enforces that the alert matches the resource’s
     namespace.

   sourceMatchers	<[]string>
     A list of matchers for which one or more alerts have to exist for the
     inhibition to take effect.

   targetMatch	<[]Object>
     Matchers that have to be fulfilled in the alerts to be muted. The operator
     enforces that the alert matches the resource’s namespace.

   targetMatchers	<[]string>
     A list of matchers that have to be fulfilled by the target alerts to be
     muted.
     
     
     
     
     
     
KIND:     AlertmanagerConfig
VERSION:  monitoring.coreos.com/v1alpha1

RESOURCE: muteTimeInterval <[]Object>

DESCRIPTION:
     List mute time intervals

     MuteTimeIntercal specifies a named interval of time that may be referenced
     in the routing tree to mute particular routes for particular times of the
     day

FIELDS:
   name	<string> -required-
     Name of a Mute Time Interval

   timeInterval	<[]Object> -required-
     Series of time intervals assoicated with the MuteTimeInterval
     



     
KIND:     AlertmanagerConfig
VERSION:  monitoring.coreos.com/v1alpha1

RESOURCE: timeInterval <[]Object>

DESCRIPTION:
     Series of time intervals assoicated with the MuteTimeInterval

     TimeInterval contains the actual definition for an interval of time.

FIELDS:
   daysOfMonth	<[]string>
     A list of numerical days in the month. Days begin at 1. Negative values are
     also accepted which begin at the end of the month

   months	<[]string>
     A list of calendar months identified by a case-insentive name (e.g.
     ‘January’) or by number, where January = 1

   times	<[]Object>
     Defines a series of start and tend times

   weekdays	<[]string>
     A list of days of the week, where the week begins on Sunday and ends on
     Saturday. Days should be specified by name (e.g. ‘Sunday’).

   years	<[]string>
     A numerical list of years. Ranges are accepted.     
```

**Release Note Template (will be copied)**

```release-note:feature
Adds support for mute_time_intervals, target_matchers and source_matchers syntax for Alertmanager
```

### CODE REVIEW

**Comment by Jammicus** [comment]
`make --always-make format generate && git diff --exit-code` is consistently failing in CI, but I can't seem to replicate locally  :(

File: pkg/alertmanager/amcfg.go [hunk 1]
```diff
@@ -121,6 +127,59 @@ func (cg *configGenerator) generateConfig(
...
+
+		x := &TimeInterval{
+			Times:       times,
+			DaysOfMonth: days,
+			Months:      months,
+			Weekdays:    week,
+			Years:       years,
+		}
+
+		// TODO, fix this hack to prevent a empty object being added to the yaml
```
Comment (fpetkovski) [inline pkg/alertmanager/amcfg.go:176]: Is this TODO something you intended address in the current PR?

File: pkg/alertmanager/amcfg.go [hunk 2]
```diff
@@ -121,6 +127,59 @@ func (cg *configGenerator) generateConfig(
 	baseConfig.Route.Routes = append(subRoutes, baseConfig.Route.Routes...)
 
 	return yaml.Marshal(baseConfig)
+
+}
+
+func convertTimeInterval(in *monitoringv1alpha1.MuteTimeInterval) *MuteTimeInterval {
+	var mti MuteTimeInterval
+
```
Comment (fpetkovski) [inline pkg/alertmanager/amcfg.go:135]:
```suggestion
```

File: pkg/alertmanager/amcfg.go [hunk 3]
```diff
@@ -121,6 +127,59 @@ func (cg *configGenerator) generateConfig(
...
+	mti.Name = in.Name
+
+	for _, interval := range in.TimeInterval {
+
```
Comment (fpetkovski) [inline pkg/alertmanager/amcfg.go:139]:
```suggestion
```

File: pkg/alertmanager/amcfg.go [hunk 4]
```diff
@@ -121,6 +127,59 @@ func (cg *configGenerator) generateConfig(
...
+		times := make([]TimeRange, len(interval.Times))
+		days := make([]string, len(interval.DaysOfMonth))
+		months := make([]string, len(interval.Months))
+		week := make([]string, len(interval.Weekdays))
+		years := make([]string, len(interval.Years))
+
+		for j, y := range interval.Times {
+
```
Comment (fpetkovski) [inline pkg/alertmanager/amcfg.go:147]:
```suggestion
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 5]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
...
 }
 
+// MuteTimeIntercal specifies a named interval of time that may be referenced
+// in the routing tree to mute particular routes for particular times of the day
+type MuteTimeInterval struct {
+	// Name of a Mute Time Interval
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+	// Series of time intervals assoicated with the MuteTimeInterval
+	TimeInterval []TimeInterval `json:"timeInterval"`
```
Comment (fpetkovski) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:732]: By convention, this variable should be named `TimeIntervals`. The json key should also be pluralized.

File: pkg/alertmanager/amcfg.go [hunk 6]
```diff
@@ -121,6 +127,59 @@ func (cg *configGenerator) generateConfig(
...
+			times[j].StartTime = y.StartTime
+			times[j].EndTime = y.EndTime
+		}
+
+		for j, y := range interval.DaysOfMonth {
```
Comment (fpetkovski) [inline pkg/alertmanager/amcfg.go:152]: Are these loops just copying the arrays? Could we use the built in copy function instead?

File: pkg/alertmanager/amcfg.go (hunk 1 above)
Comment (Jammicus) [reply pkg/alertmanager/amcfg.go:176]: Updated comment.

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go (hunk 5 above)
Comment (Jammicus) [reply pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:732]: Ahhhh awesome didn't know this, will update

File: pkg/alertmanager/amcfg.go (hunk 6 above)
Comment (Jammicus) [reply pkg/alertmanager/amcfg.go:152]: Good idea, will update

**Review by fpetkovski — CHANGES_REQUESTED** [review]
Hey @Jammicus, thanks for the follow up changes. 

One thing which I missed initially was that we also need to address backwards compatibility. Running Alertmanager 0.21 with the newly added config options will result in the following error 
```
level=error ts=2021-07-14T06:47:14.211Z caller=coordinator.go:124 component=configuration msg="Loading configuration file failed" file=/etc/alertmanager/config/alertmanager.yaml err="yaml: unmarshal errors:\n  line 14: field mute_time_intervals not found in type config.plain"
```

What we usually do in this cases is render the new config only for versions which support it. You can see one such example here: https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/prometheus/promcfg.go#L206

This PR needs to do a similar thing where the newly added configuration options would only be applied for AM versions 0.22 and above.

**Comment by Jammicus** [comment]
> Hey @Jammicus, thanks for the follow up changes.
> 
> One thing which I missed initially was that we also need to address is backwards compatibility. Running Alertmanager 0.21 with the newly added config options will result in the following error
> 
> ```
> level=error ts=2021-07-14T06:47:14.211Z caller=coordinator.go:124 component=configuration msg="Loading configuration file failed" file=/etc/alertmanager/config/alertmanager.yaml err="yaml: unmarshal errors:\n  line 14: field mute_time_intervals not found in type config.plain"
> ```
> 
> What we usually do in this cases is render the new config only for versions which support it. You can see one such example here: https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/prometheus/promcfg.go#L206
> 
> This PR needs to do a similar thing where the newly added configuration options would only be applied for AM versions 0.22 and above.

Super thanks for this, will try to do the changes this weekend

File: pkg/alertmanager/amcfg_test.go [hunk 7]
```diff
@@ -23,6 +23,7 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/prometheus-operator/prometheus-operator/pkg/assets"
 	"github.com/prometheus/alertmanager/config"
+
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg_test.go:26]: not needed ;)

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 8]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
 	Regex bool `json:"regex,omitempty"`
 }
 
+// MuteTimeIntercal specifies a named interval of time that may be referenced
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:725]:
```suggestion
// MuteTimeInterval specifies a named interval of time that may be referenced
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 9]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
...
+// in the routing tree to mute particular routes for particular times of the day
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:726]:
```suggestion
// in the routing tree to mute particular routes for particular times of the day.
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 10]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
...
+type MuteTimeInterval struct {
+	// Name of a Mute Time Interval
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:728]:
```suggestion
	// Name of the mute time interval.
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 11]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
...
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+	// Series of time intervals assoicated with the MuteTimeInterval
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:731]:
```suggestion
	// Series of time intervals associated with the mute time interval.
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 12]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
...
+	TimeIntervals []TimeInterval `json:"timeIntervals"`
+}
+
+// TimeInterval  contains the actual definition for an interval of time.
+type TimeInterval struct {
+	// Defines a series of start and tend times
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:737]:
```suggestion
	// Defines a series of start and end times.
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 13]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
...
+	Times []TimeRange `json:"times,omitempty"`
+	// A list of days of the week, where the week begins on Sunday and ends on Saturday.
+	//Days should be specified by name (e.g. ‘Sunday’).
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:740]:
```suggestion
	// Days should be specified by name (e.g. ‘Sunday’).
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 14]
```diff
@@ -713,6 +722,43 @@ type Matcher struct {
...
+	// +optional
+	Weekdays []string `json:"weekdays,omitempty"`
+	// A list of numerical days in the month. Days begin at 1. Negative values are also accepted which begin at the end of the month
+	// +optional
+	DaysOfMonth []string `json:"daysOfMonth,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:745]: shouldn't it be DaysOfMonthRange?

**Review by simonpasquier — COMMENTED** [review]
Thanks a lot @Jammicus for working on it, this looks very good so far! As @fpetkovski already said, the operator needs to support the new fields based on the Alertmanager version being deployed (and degrade gracefully if the version doesn't support the new fields).
I would also suggest that you split this PR into (at least) 2 PRs: one for supporting the mute time intervals and another one for target_matchers/source_matchers (which also applies to routes). AFAICT the 2 changes aren't directly related and it would be easier for us to review (probably also easier for you to deal with the feedback).

**Comment by Jammicus** [comment]
> Thanks a lot @Jammicus for working on it, this looks very good so far! As @fpetkovski already said, the operator needs to support the new fields based on the Alertmanager version being deployed (and degrade gracefully if the version doesn't support the new fields).
> I would also suggest that you split this PR into (at least) 2 PRs: one for supporting the mute time intervals and another one for target_matchers/source_matchers (which also applies to routes). AFAICT the 2 changes aren't directly related and it would be easier for us to review (probably also easier for you to deal with the feedback).

Thanks for this - will split out the the target_matchers and source_matchers into a separate PR

**Comment by Jammicus** [comment]
Closing this in favour of two split out PRs referenced in the first post.

**Pull request closed without merging**